### PR TITLE
Instanced mesh example

### PR DESCRIPTION
**Pull request by brisvag**
A surprisingly simple visual for arbitrarily oriented and colored meshes using instanced rendering.

I started on this as a way to have "true 3D" markers for solving #2453, but adding arbitrary orientations turned out to be relatively easy, and that's useful fo r stuff I'm working on :P

Basically, replace the mesh with a simple [icosphere](https://github.com/vedranaa/icosphere/) and you get your real pysically sized markers that work in both 2D and 3D. Notice how I applied a scaling transform to the visual in the examples, and the dinosaurs are being stretched correctly.

https://user-images.githubusercontent.com/23482191/224005681-3297902e-fd78-4fdf-a782-a14dac41d936.mp4

One thing that's missing (and likely hard to do) compared to the markers is the edge. As far as I know, the only 2 reasonable ways to do it are geometry shaders and multipass rendering?

### CODE REVIEW

**Comment by alisterburt** [comment]
:heart: :heart: :heart:

File: examples/scene/instanced_mesh.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+view.camera = 'arcball'
+view.camera.scale_factor = 1000
+
+N = 1000
+
+mesh_file = load_data_file('orig/triceratops.obj.gz')
+vertices, faces, _, _ = read_mesh(mesh_file)
+
+np.random.seed(0)
+pos = (np.random.rand(N, 3) - 0.5) * 1000
```
Comment (kephale) [inline examples/scene/instanced_mesh.py:104]: 1000 should be N

**Comment by brisvag** [comment]
@djhoese this would be nice to get in, but I don't know how to intepret that test failure. Something about discontinuous data being copied? How can I solve it?

**Comment by djhoese** [comment]
Hm and you don't see this warning when you run it locally? Same version of python and numpy? Any idea if the warning is **in** vispy?

File: examples/scene/instanced_mesh.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        # of instances
+        self.shifts = gloo.VertexBuffer(positions.astype(np.float32), divisor=1)
+        self.shared_program['shift'] = self.shifts
+
+        # vispy does not handle matrix attributes (likely requires some big changes in GLIR)
+        # so we decompose it into three vec3; (column vectors of the matrix)
+        transforms = transforms.astype(np.float32)
+        self.transforms_x = gloo.VertexBuffer(transforms[..., 0], divisor=1)
+        self.transforms_y = gloo.VertexBuffer(transforms[..., 1], divisor=1)
+        self.transforms_z = gloo.VertexBuffer(transforms[..., 2], divisor=1)
```
Comment (djhoese) [inline examples/scene/instanced_mesh.py:74]: My guess is this is the problem regarding the dis-contiguous data. By doing the indexing there you're now causing striding by 3 (for the 3 elements X/Y/Z) and the VertexBuffer underneath is complaining that the data isn't contiguous and has to be copied.

https://github.com/vispy/vispy/blob/ded8eb9d7483255789dcac87b01399bd3925905b/vispy/gloo/buffer.py#L438-L442

Comment (djhoese) [reply examples/scene/instanced_mesh.py:74]: So I think if you did a `.copy()` here when you do the `[..., 0]` style indexing then you'd at least avoid the warning. If you could find a way to structure the array as `(3, M, N)` then do indexing `[0]` then you'd also avoid the copy.

**Comment by djhoese** [comment]
Ok, passing now. Anything else before I do one last skim of the code and merge?

**Comment by brisvag** [comment]
Nope, I think it's good to go :)

File: examples/scene/instanced_mesh.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
+# -*- coding: utf-8 -*-
+# -----------------------------------------------------------------------------
+# Copyright (c) Vispy Development Team. All Rights Reserved.
+# Distributed under the (new) BSD License. See LICENSE.txt for more info.
+# -----------------------------------------------------------------------------
+"""
+Instanced rendering of arbitrarily transformed meshes
+=====================================================
+"""
```
Comment (djhoese) [inline examples/scene/instanced_mesh.py:9]: It looks like this example isn't getting a proper thumbnail/image extracted from it. Do you want to try adding a `gallery` line and see if the doc rendering can generate an image?

https://vispy.org/dev_guide/writing_examples.html#specify-frames-to-grab

**Comment by djhoese** [comment]
I downloaded the generated `.zip` of the website and the image looks good. Thanks. Merging...

**Pull request merged**